### PR TITLE
ABI/codecs implementation (part 1)

### PR DESCRIPTION
**Pull request by andreibancioiu**
This is part of a series of pull requests.

See: https://github.com/multiversx/mx-sdk-py/pull/32.

This can be reviewed side-by-side with:
 - https://github.com/multiversx/mx-sdk-abi-go/blob/main/abi/serializer.go
 - https://github.com/multiversx/mx-sdk-abi-go/blob/main/abi/serializer_test.go
 - https://github.com/multiversx/mx-sdk-abi-go/blob/main/abi/codec.go

### CODE REVIEW

File: multiversx_sdk/abi/serializer.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    def serialize_to_parts(self, input_values: Sequence[Any]) -> List[bytes]:
+        parts_holder = PartsHolder([])
+        self._do_serialize(parts_holder, input_values)
+        return parts_holder.get_parts()
+
+    def _do_serialize(self, parts_holder: PartsHolder, input_values: Sequence[Any]):
+        for i, value in enumerate(input_values):
+            if value is None:
+                raise ValueError("cannot serialize nil value")
```
Comment (popenta) [inline multiversx_sdk/abi/serializer.py:33]: `nil` seems very Golang-ish

File: multiversx_sdk/abi/serializer.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        self.deserialize_parts(parts, output_values)
+
+    def deserialize_parts(self, parts: Sequence[bytes], output_values: Sequence[Any]):
+        parts_holder = PartsHolder(parts)
+        self._do_deserialize(parts_holder, output_values)
+
+    def _do_deserialize(self, parts_holder: PartsHolder, output_values: Sequence[Any]):
+        for i, value in enumerate(output_values):
+            if value is None:
+                raise ValueError("cannot deserialize into nil value")
```
Comment (popenta) [inline multiversx_sdk/abi/serializer.py:70]: same comm as above regarding `nil`

File: multiversx_sdk/abi/serializer.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+from multiversx_sdk.abi.values_multi import *
+
+
+class Serializer:
+    """
+    The serializer follows the rules of the MultiversX Serialization format:
+    https://docs.multiversx.com/developers/data/serialization-overview
+    """
+
+    def __init__(self, parts_separator: str):
```
Comment (popenta) [inline multiversx_sdk/abi/serializer.py:14]: maybe `parts_separator` can be `@` by default? But i also understand this respects the Go implementation. Not a big deal anyway.

**Review by popenta — APPROVED** [review]

**Review by axenteoctavian — APPROVED** [review]

File: multiversx_sdk/abi/serializer.py (hunk 1 above)
Comment (andreibancioiu) [reply multiversx_sdk/abi/serializer.py:33]: Fixed in https://github.com/multiversx/mx-sdk-py/pull/49.

File: multiversx_sdk/abi/serializer.py (hunk 2 above)
Comment (andreibancioiu) [reply multiversx_sdk/abi/serializer.py:70]: Fixed in https://github.com/multiversx/mx-sdk-py/pull/49.

File: multiversx_sdk/abi/serializer.py (hunk 3 above)
Comment (andreibancioiu) [reply multiversx_sdk/abi/serializer.py:14]: The `@` separator is provided in the higher-level components. For example: 
 - https://github.com/multiversx/mx-sdk-py/blob/codecs-init/multiversx_sdk/core/transactions_factories/smart_contract_transactions_factory.py#L45

**Pull request merged**